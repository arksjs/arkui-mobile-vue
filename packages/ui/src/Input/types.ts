import type { FormItemCommonProps, FocusEmits } from '../Form/types'

export type Mode = 'search' | 'numeric' | 'decimal' | 'tel' | 'text' | 'none'

type OnChange = (value: string) => void

export interface InputProps extends FormItemCommonProps {
  maxlength?: number | string
  placeholder?: string
  type?: string
  value?: string
  focus?: boolean
  readonly?: boolean
  showClear?: boolean
  showLimit?: boolean
}

export interface InputEmits extends FocusEmits {
  onInput?: OnChange
  onChange?: OnChange
}
