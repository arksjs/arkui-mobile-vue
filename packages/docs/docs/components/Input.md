# Input 输入框

注：

- 支持表单，具体可参考 [Form](./Form.md)。

<CodeDemo name="Input">

<<< @/../../demo/src/components/Form/Input/index.vue

</CodeDemo>

## Import

```js
import { FxInput } from 'arkui-mobile-vue'
```

具体的引入方式可以参考[引入组件](../guide/import.md)。

## Props

| 属性        | 类型             | 默认值 | 必填 | 说明               |
| ----------- | ---------------- | ------ | ---- | ------------------ |
| v-model     | string \| number | ''     | 否   |
| name        | string           |        | 否   | 标识               |
| type        | string           | 'text' | 否   | 类型               |
| placeholder | string           |        | 否   | 输入框为空时占位符 |
| disabled    | boolean          | false  | 否   | 是否禁用           |
| readonly    | boolean          | false  | 否   | 是否只读           |
| maxlength   | string \| number | 140    | 否   | 最大长度           |
| focus       | boolean          | false  | 否   | 是否获取焦点       |
| show-clear  | boolean          | false  | 否   | 是否展示清除图标   |

### type 的合法值

| 值       | 说明 |
| -------- | ---- |
| text     | 文本 |
| digit    | 整数 |
| number   | 数字 |
| password | 密码 |
| search   | 搜索 |
| password | 密码 |

## Events

| 事件   | 描述                                   | 回调函数参数  |
| ------ | -------------------------------------- | ------------- |
| input  | 输入值改变时触发                       | value: string |
| change | 输入值改变且失焦后触发，含点击清空按钮 | value: string |
| focus  | 获取焦点时触发                         | e: FocusEvent |
| blur   | 移出焦点时触发                         | e: FocusEvent |

## Slots

### 前置元素（#prepend）

```vue
<ak-input>
  <template #prepend>https://</template>
</ak-input>
```

注：也可以传入 `Icon`，比如常见的搜索。

### 后置元素（#append）

```vue
<ak-input>
  <template #append>.com</template>
</ak-input>
```
