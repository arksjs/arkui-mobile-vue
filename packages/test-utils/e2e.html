<script src="../../node_modules/vue/dist/vue.global.js"></script>
<script src="../../dist/index.js"></script>
<link rel="stylesheet" href="../../dist/index.css" />

<div id="app"></div>
