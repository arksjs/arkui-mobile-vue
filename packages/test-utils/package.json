{
  "name": "@arksjs/test-utils",
  "version": "0.0.1",
  "description": "arkui test-utils",
  "main": "dist/index.js",
  "private": true
}
