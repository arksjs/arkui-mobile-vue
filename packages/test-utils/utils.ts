export const timeout = (n: number) => new Promise(r => setTimeout(r, n))
