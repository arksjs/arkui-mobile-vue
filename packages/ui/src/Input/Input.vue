<template>
  <label :class="classes">
    <div class="ak-input_prepend" v-if="$slots.prepend">
      <slot name="prepend"></slot>
    </div>
    <textarea
      v-if="type === 'textarea'"
      class="ak-input_input ak-input_textarea"
      :name="name"
      :disabled="disabled"
      :placeholder="placeholder"
      :readonly="readonly"
      :maxlength="maxLength"
      @input="onInput"
      @change="onChange"
      @focus="onFocus"
      @blur="onBlur"
      ref="inputEl"
    ></textarea>
    <input
      v-else
      class="ak-input_input"
      :name="name"
      :type="typeMode.inputType"
      :inputmode="typeMode.inputMode"
      :disabled="disabled"
      :placeholder="placeholder"
      :readonly="readonly"
      :maxlength="maxLength"
      @input="onInput"
      @change="onChange"
      @focus="onFocus"
      @blur="onBlur"
      @compositionstart="onCompositionStart"
      @compositionend="onCompositionEnd"
      ref="inputEl"
    />
    <span class="ak-input_limit" v-if="showLimit && maxLength > 0"
      >{{ inputValue.length }}/{{ maxLength }}</span
    >
    <Icon
      v-if="showClear && isShowClear"
      class="ak-input_clear"
      :icon="CloseCircleFilled"
      @mousedown.prevent="onClear"
    />
    <div class="ak-input_append" v-if="$slots.append">
      <slot name="append"></slot>
    </div>
  </label>
</template>

<script lang="ts">
import { computed, defineComponent, onMounted, ref, watch } from 'vue'
import { Icon } from '../Icon'
import { isNumeric, isStringNumberMix } from '../helpers/util'
import { getClasses, getInputMode, getMaxLength, getValue } from './util'
import {
  formFocusEmits,
  formItemProps,
  formStringValueEmits
} from '../Form/form'
import { useInput } from '../Form/use-form'
import CloseCircleFilled from '../Icon/icons/CloseCircleFilled'
import type { OnFocus, PropsToEmits } from '../helpers/types'
import type { InputEmits } from './types'

export default defineComponent({
  name: 'ak-input',
  components: { Icon },
  props: {
    ...formItemProps,
    maxlength: {
      type: [Number, String],
      validator: (val: number | string) => isNumeric(val),
      default: 140
    },
    placeholder: {
      type: String,
      default: ''
    },
    type: {
      type: String,
      default: 'text'
    },
    modelValue: {
      type: [Number, String],
      validator: (val: number | string) => isStringNumberMix(val)
    },
    focus: {
      type: Boolean,
      default: false
    },
    readonly: {
      type: Boolean,
      default: false
    },
    showClear: {
      type: Boolean,
      default: false
    },
    // 展示字数限制文本
    showLimit: {
      type: Boolean,
      default: false
    }
  },
  emits: {
    ...formStringValueEmits,
    ...formFocusEmits
  } as PropsToEmits<InputEmits>,
  setup(props, { emit, slots }) {
    const active = ref(false)
    const isShowClear = ref(false)
    const inputValue = ref('')

    const { inputEl, setFocus, setBlur, getInputValue, setInputValue } =
      useInput()

    function updateValue(val: string | number) {
      const newVal = getValue(val, props.type)

      let isChange = false

      if (newVal !== getInputValue()) {
        setInputValue(newVal)
      }

      if (newVal !== inputValue.value) {
        inputValue.value = newVal
        isChange = true
      }

      if (newVal != props.modelValue) {
        emit('update:modelValue', newVal)
      }

      return { value: newVal, isChange }
    }

    function updateInput(newVal: string) {
      const { value, isChange } = updateValue(newVal)

      isChange && emit('input', value)
    }

    let isComposition = false

    function onCompositionStart() {
      isComposition = true
    }

    function onCompositionEnd() {
      isComposition = false
      updateInput(getInputValue())
    }

    const onFocus: OnFocus = e => {
      active.value = true
      emit('focus', e)
    }
    const onBlur: OnFocus = e => {
      active.value = false
      emit('blur', e)
    }

    function onInput() {
      if (!isComposition) {
        updateInput(getInputValue())
      }
    }

    function onChange() {
      emit('change', inputValue.value)
    }

    function onClear() {
      updateInput('')
      emit('change', inputValue.value)
    }

    const typeMode = computed(() => getInputMode(props.type))
    const classes = computed(() =>
      getClasses({
        type: props.type,
        prepend: !!slots.prepend,
        append: !!slots.append,
        active: active.value,
        disabled: props.disabled,
        readonly: props.readonly
      })
    )

    watch(
      () => props.modelValue,
      val => {
        val != inputValue.value && updateValue(val ?? '')
      }
    )

    watch(
      () => props.focus,
      val => {
        val ? setFocus() : setBlur()
      }
    )

    const maxLength = computed(() => getMaxLength(props.maxlength))

    let timer: number
    watch([inputValue, active], ([ipVal, isActive]) => {
      clearTimeout(timer)
      if (ipVal && isActive) {
        timer = window.setTimeout(() => (isShowClear.value = true), 200)
      } else {
        isShowClear.value = false
      }
    })

    onMounted(() => {
      updateValue(props.modelValue ?? '')

      props.focus && setFocus()
    })

    return {
      classes,
      inputEl,
      inputValue,
      active,
      onCompositionStart,
      onCompositionEnd,
      onFocus,
      onBlur,
      onInput,
      onChange,
      onClear,
      typeMode,
      maxLength,
      isShowClear,
      CloseCircleFilled
    }
  }
})
</script>
