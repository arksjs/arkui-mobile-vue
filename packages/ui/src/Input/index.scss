@import '../style/var.scss';

.ta-input {
  --ta-font-size: 17px;
  --ta-input-icon-size: 20px;
  --ta-front-color: #{$title-color};
  --ta-secondary-color: #{$font3-color};

  height: 40px;
  width: 100%;
  position: relative;
  display: flex;
  align-items: center;
  border: 1px solid $border-color;
  border-radius: 3px;
  box-sizing: border-box;
  padding: 7px 12px;
  font-size: var(--ta-font-size);
  background-color: $white-color;
  color: var(--ta-front-color);

  .ta-icon {
    --ta-icon-size: var(--ta-input-icon-size);
    --ta-icon-color: var(--ta-secondary-color);
  }

  &.disabled {
    background-color: $background-color;
    opacity: 0.2;
    cursor: not-allowed;
  }

  &_prepend,
  &_append {
    display: flex;
    align-items: center;
    padding: 0 8px 0 0;
    color: var(--ta-secondary-color);
  }

  &_append,
  &_clear,
  &_arrow {
    padding: 0 0 0 8px;
  }

  &.has--value {
    .ta-input {
      &_arrow {
        --ta-icon-color: var(--ta-front-color);
      }
    }
  }

  &_limit {
    font-size: 15px;
    line-height: 24px;
    height: 24px;
    color: $font-color;
    margin-left: 8px;
  }

  &_input {
    flex: 1;
    display: block;
    margin: 0;
    outline: none;
    border: none;
    width: 100%;
    height: 100%;
    line-height: 1.412;
    padding: 0;
    font-size: var(--ta-font-size);
    cursor: pointer;
    color: var(--ta-front-color);
    background: none;
    box-sizing: border-box;
    box-shadow: none;
    resize: none;

    &.ellipsis {
      display: -webkit-box;
      -webkit-line-clamp: 1;
      -webkit-box-orient: vertical;
      overflow: hidden;
    }

    // white-space: nowrap;
    // text-overflow: ellipsis;
    // overflow: hidden;

    &:read-only {
      cursor: auto;
    }

    &[type='search']::-webkit-search-cancel-button {
      display: none;
    }

    &.placeholder,
    &::placeholder {
      color: $font3-color;
    }

    &:disabled {
      user-select: none;
    }
  }

  &:not(.readonly).focus {
    border-color: $primary-color;
  }

  &_cover {
    position: absolute;
    left: 0;
    top: 0;
    width: 100%;
    height: 100%;
    opacity: 0;
    margin: 0;
    padding: 0;
    border: none;

    &:disabled {
      cursor: not-allowed;
    }
  }
}

.ta-textarea {
  height: 108px;

  .ta-input_limit {
    position: absolute;
    right: 12px;
    bottom: 8px;
  }
}
