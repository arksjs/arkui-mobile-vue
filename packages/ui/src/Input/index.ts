import { withInstall } from '../helpers/with-install'
import Input from './Input.vue'

export { Input }
export const install = withInstall(Input)
export default Input
