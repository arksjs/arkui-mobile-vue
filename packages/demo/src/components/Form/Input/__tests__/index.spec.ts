import { shallowMount } from '@vue/test-utils'
import { Input } from '@/Input'

describe('Input', () => {
  test('should render default correctly', () => {
    const wrapper = shallowMount(Input, {})

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('should render prop showClear=true correctly', () => {
    const wrapper = shallowMount(Input, {
      props: {
        showClear: true
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('should render prop maxlength=200 correctly', () => {
    const wrapper = shallowMount(Input, {
      props: {
        maxlength: 200
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('should render prop disabled=true correctly', () => {
    const wrapper = shallowMount(Input, {
      props: {
        disabled: true
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
