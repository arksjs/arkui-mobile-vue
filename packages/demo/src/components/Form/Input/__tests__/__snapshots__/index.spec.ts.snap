// Vitest Snapshot v1

exports[`Input > should render default correctly 1`] = `
"<label class=\\"fx-input\\">
  <!--v-if--><input class=\\"fx-input_input\\" name=\\"\\" type=\\"text\\" inputmode=\\"text\\" placeholder=\\"\\" maxlength=\\"140\\">
  <!--v-if-->
  <!--v-if-->
  <!--v-if-->
</label>"
`;

exports[`Input > should render prop disabled=true correctly 1`] = `
"<label class=\\"fx-input disabled\\">
  <!--v-if--><input class=\\"fx-input_input\\" name=\\"\\" type=\\"text\\" inputmode=\\"text\\" disabled=\\"\\" placeholder=\\"\\" maxlength=\\"140\\">
  <!--v-if-->
  <!--v-if-->
  <!--v-if-->
</label>"
`;

exports[`Input > should render prop maxlength=200 correctly 1`] = `
"<label class=\\"fx-input\\">
  <!--v-if--><input class=\\"fx-input_input\\" name=\\"\\" type=\\"text\\" inputmode=\\"text\\" placeholder=\\"\\" maxlength=\\"200\\">
  <!--v-if-->
  <!--v-if-->
  <!--v-if-->
</label>"
`;

exports[`Input > should render prop showClear=true correctly 1`] = `
"<label class=\\"fx-input\\">
  <!--v-if--><input class=\\"fx-input_input\\" name=\\"\\" type=\\"text\\" inputmode=\\"text\\" placeholder=\\"\\" maxlength=\\"140\\">
  <!--v-if-->
  <!--v-if-->
  <!--v-if-->
</label>"
`;
